### PR TITLE
Add devoxelize Op

### PR DESCRIPTION
**Pull request by sanskar107**
Added `trilinear_devoxelize` and its gradient Op for Point Voxel Convolution.

---
This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d/4003)

### CODE REVIEW

**Review by benjaminum — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4003)* status: 0 of 11 files reviewed, 3 unresolved discussions (waiting on @sanskar107)

___
*[cpp/open3d/ml/pytorch/pvcnn/TrilinearDevoxelizeKernel.h, line 29 at r2](https://reviewable.io/reviews/isl-org/open3d/4003#-Mja14Pi5M2h9s9mdu45:-Mja14Pi5M2h9s9mdu46:b-t5frpy) ([raw file](https://github.com/isl-org/open3d/blob/a14b6dade337cc2c372e59173b0842175cbd7921/cpp/open3d/ml/pytorch/pvcnn/TrilinearDevoxelizeKernel.h#L29)):*
> ```objc
> #pragma once
> 
> void trilinear_devoxelize(int b,
> ```

Please add documentation
___
*[cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp, line 40 at r2](https://reviewable.io/reviews/isl-org/open3d/4003#-Mja2kwF9n_8uC-DwlND:-Mja2kwF9n_8uC-DwlNE:btfl000) ([raw file](https://github.com/isl-org/open3d/blob/a14b6dade337cc2c372e59173b0842175cbd7921/cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp#L40)):*
> ```cpp
>         .Input("features: float32")
>         .Output("outs: float32")
>         .Output("inds: int32")
> ```

Let's use the full names for the variables.
___
*[cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp, line 42 at r2](https://reviewable.io/reviews/isl-org/open3d/4003#-Mja2FV8846zUEya3ylU:-Mja2FV98TbETn_ZU8d4:b-me2vvu) ([raw file](https://github.com/isl-org/open3d/blob/a14b6dade337cc2c372e59173b0842175cbd7921/cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp#L42)):*
> ```cpp
>         .Output("inds: int32")
>         .Output("wgts: float32")
>         .SetShapeFn([](::tensorflow::shape_inference::InferenceContext* c) {
> ```

Consider using macros from `TensorflowHelper.h` and `TorchHelper.h` to make shape checking easier.

**Review by sanskar107 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4003)* status: 0 of 11 files reviewed, 3 unresolved discussions (waiting on @benjaminum)

___
*[cpp/open3d/ml/pytorch/pvcnn/TrilinearDevoxelizeKernel.h, line 29 at r2](https://reviewable.io/reviews/isl-org/open3d/4003#-Mja14Pi5M2h9s9mdu45:-MjhumKo-d3NfC1wrWEW:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/a14b6dade337cc2c372e59173b0842175cbd7921/cpp/open3d/ml/pytorch/pvcnn/TrilinearDevoxelizeKernel.h#L29)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Please add documentation
</blockquote></details>

Done.
___
*[cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp, line 40 at r2](https://reviewable.io/reviews/isl-org/open3d/4003#-Mja2kwF9n_8uC-DwlND:-MjhuoKNBVBSz2pVdzsn:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/a14b6dade337cc2c372e59173b0842175cbd7921/cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp#L40)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Let's use the full names for the variables.
</blockquote></details>

Done.
___
*[cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp, line 42 at r2](https://reviewable.io/reviews/isl-org/open3d/4003#-Mja2FV8846zUEya3ylU:-Mjhuqcf06bNgoQFV_gV:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/a14b6dade337cc2c372e59173b0842175cbd7921/cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp#L42)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Consider using macros from `TensorflowHelper.h` and `TorchHelper.h` to make shape checking easier.
</blockquote></details>

Done.

**Review by benjaminum — CHANGES_REQUESTED** [review]
How does the graph building work for the gradient? I don't see the code relating the forward and backward ops.

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4003#-:-MjnkgfTE7G4ooByvnCc:b-m3fyap)* status: 0 of 11 files reviewed, 7 unresolved discussions (waiting on @benjaminum and @sanskar107)

___
*[cpp/open3d/ml/contrib/cuda_utils.h, line 40 at r3](https://reviewable.io/reviews/isl-org/open3d/4003#-MjnkOW8ERzl8X8gifHE:-MjnkOW8ERzl8X8gifHF:b-8hfykq) ([raw file](https://github.com/isl-org/open3d/blob/95a6a9aa36d5e16e927aad0c033ff7054fa53c36/cpp/open3d/ml/contrib/cuda_utils.h#L40)):*
> ```objc
> #define DIVUP(m, n) ((m) / (n) + ((m) % (n) > 0))
> 
> inline int opt_n_threads(int work_size) {
> ```

CamelCase for function names
___
*[cpp/open3d/ml/contrib/TrilinearDevoxelize.cu, line 33 at r3](https://reviewable.io/reviews/isl-org/open3d/4003#-MjnkKRp-gUMLf3PuCx7:-MjnkKRp-gUMLf3PuCx8:b-5p0uvg) ([raw file](https://github.com/isl-org/open3d/blob/95a6a9aa36d5e16e927aad0c033ff7054fa53c36/cpp/open3d/ml/contrib/TrilinearDevoxelize.cu#L33)):*
> ```Cuda
> namespace contrib {
> 
> __global__ void trilinear_devoxelize_kernel(int b,
> ```

CamelCase same for all other functions.
___
*[cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp, line 75 at r3](https://reviewable.io/reviews/isl-org/open3d/4003#-Mjnjxc_DuHkcrmZgMqU:-Mjnjxc_DuHkcrmZgMqV:bpk0dq1) ([raw file](https://github.com/isl-org/open3d/blob/95a6a9aa36d5e16e927aad0c033ff7054fa53c36/cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp#L75)):*
> ```cpp
>             return Status::OK();
>         })
>         .Doc(R"doc( TODO )doc");
> ```

Please add docstring
___
*[cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp, line 109 at r3](https://reviewable.io/reviews/isl-org/open3d/4003#-Mjnk-9s1xr4EOFfkRDg:-Mjnk-9s1xr4EOFfkRDh:bqljqib) ([raw file](https://github.com/isl-org/open3d/blob/95a6a9aa36d5e16e927aad0c033ff7054fa53c36/cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp#L109)):*
> ```cpp
>             return Status::OK();
>         })
>         .Doc(R"doc( TODO )doc");
> ```

This one can be short if the gradient op is not used directly by the users

**Review by benjaminum — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4003)* status: 0 of 13 files reviewed, 8 unresolved discussions (waiting on @benjaminum and @sanskar107)

___
*[cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp, line 140 at r4](https://reviewable.io/reviews/isl-org/open3d/4003#-MkHfqMp1SxTBtpsXxg8:-MkHfqMp1SxTBtpsXxg9:b-qkt7el) ([raw file](https://github.com/isl-org/open3d/blob/b4b5223b963278004e3b8def27664633cb1bad5d/cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp#L140)):*
> ```cpp
> 
> coords: List of 3D coordinates for which features to be interpolated.
>   The shape of this tensor is [B, 3, N].
> ```

What is the range of the coordinates and where is the voxel center? What happens if the coordinate is outside of the range?

**Review by sanskar107 — COMMENTED** [review]
Graph building is implemented in python in Open3D-ML. https://github.com/isl-org/Open3D-ML/blob/sanskar/pvcnn/ml3d/torch/models/utils.py#L10

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4003#-:-MkN7fOzCJlpAN6qmekw:boh7zz1)* status: 0 of 13 files reviewed, 8 unresolved discussions (waiting on @benjaminum)

___
*[cpp/open3d/ml/contrib/cuda_utils.h, line 40 at r3](https://reviewable.io/reviews/isl-org/open3d/4003#-MjnkOW8ERzl8X8gifHE:-MkN7uCP7Sqo8gm7sv0F:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/95a6a9aa36d5e16e927aad0c033ff7054fa53c36/cpp/open3d/ml/contrib/cuda_utils.h#L40)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

CamelCase for function names
</blockquote></details>

Done.
___
*[cpp/open3d/ml/contrib/TrilinearDevoxelize.cu, line 33 at r3](https://reviewable.io/reviews/isl-org/open3d/4003#-MjnkKRp-gUMLf3PuCx7:-MkN7uxB4HHTxwhnHFeT:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/95a6a9aa36d5e16e927aad0c033ff7054fa53c36/cpp/open3d/ml/contrib/TrilinearDevoxelize.cu#L33)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

CamelCase same for all other functions.
</blockquote></details>

Done.
___
*[cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp, line 75 at r3](https://reviewable.io/reviews/isl-org/open3d/4003#-Mjnjxc_DuHkcrmZgMqU:-MkN7vt2EOSqk46R5C9u:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/95a6a9aa36d5e16e927aad0c033ff7054fa53c36/cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp#L75)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Please add docstring
</blockquote></details>

Done.
___
*[cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp, line 109 at r3](https://reviewable.io/reviews/isl-org/open3d/4003#-Mjnk-9s1xr4EOFfkRDg:-MkN7xB7BAnzfGxKn0sb:b-896fix) ([raw file](https://github.com/isl-org/open3d/blob/95a6a9aa36d5e16e927aad0c033ff7054fa53c36/cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp#L109)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

This one can be short if the gradient op is not used directly by the users
</blockquote></details>

Done.
___
*[cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp, line 140 at r4](https://reviewable.io/reviews/isl-org/open3d/4003#-MkHfqMp1SxTBtpsXxg8:-MkN8-JGBOmuc-qqdKhb:bfop3oz) ([raw file](https://github.com/isl-org/open3d/blob/b4b5223b963278004e3b8def27664633cb1bad5d/cpp/open3d/ml/tensorflow/pvcnn/TrilinearDevoxelizeOps.cpp#L140)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

What is the range of the coordinates and where is the voxel center? What happens if the coordinate is outside of the range?
</blockquote></details>

Added to docs.

**Review by benjaminum — COMMENTED** [review]
@sanskar107 We can merge after fixing the style checks

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4003#-:-Mk_oQf48htfymA6KSwM:b-36n1z0)* status: 0 of 13 files reviewed, 8 unresolved discussions (waiting on @benjaminum)

**Review by benjaminum — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d/4003#-:-MkpHo7B23fBIZHqzt-n:bnfp4nl)* status: 0 of 13 files reviewed, 8 unresolved discussions (waiting on @benjaminum)
